### PR TITLE
Enable mypy option `--disallow-any-generics`

### PR DESCRIPTION
**Pull request by michaeloliverx**
Refs https://github.com/encode/httpx/issues/2436

### CODE REVIEW

File: httpx/_transports/wsgi.py [hunk 1]
```diff
@@ -7,8 +7,14 @@
 from .._types import SyncByteStream
 from .base import BaseTransport
 
+if typing.TYPE_CHECKING:
+    from _typeshed import OptExcInfo  # pragma: no cover
+    from _typeshed.wsgi import WSGIApplication  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (g-as) [reply httpx/_transports/wsgi.py:12]: https://github.com/python/typeshed/tree/main/stdlib/_typeshed#usage

[the remark this answers is not in the document]
Comment (g-as) [reply httpx/_transports/wsgi.py:12]: > The _typeshed package and its types do not exist at runtime, but can be used freely in stubs (.pyi) files. To import the types from this package in implementation (.py) files, use the following construct:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from _typeshed import ...
```

File: httpx/_auth.py [hunk 2]
```diff
@@ -10,6 +10,9 @@
 from ._models import Request, Response
 from ._utils import to_bytes, to_str, unquote
 
+if typing.TYPE_CHECKING:  # pragma: no cover
+    from hashlib import _Hash
```
[the remark this answers is not in the document]
Comment (michaeloliverx) [reply httpx/_auth.py]: I resolved it by using a protocol instead.

File: httpx/_transports/wsgi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaeloliverx) [reply httpx/_transports/wsgi.py:12]: These are utility types used by the typeshed, I see no advantage in replicating them ourselves as mypy and pyright both ship with a recent copy of typeshed. These are strictly used for type checking and have no runtime implications. 

Read more at https://github.com/python/typeshed/tree/main/stdlib/_typeshed.

**Comment by michaeloliverx** [comment]
@tomchristie sorry I was out for a bit, this should be good to go now.

File: httpx/_transports/wsgi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaeloliverx) [reply httpx/_transports/wsgi.py:12]: These particular types exist in python 3.11 under [`wsgiref.types`](https://docs.python.org/3/library/wsgiref.html#module-wsgiref.types) so eventually we can remove them.

**Comment by michaeloliverx** [comment]
This has been covered by https://github.com/encode/httpx/pull/2468

**Pull request closed without merging**